src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java

```diff
@@ -183,5 +192,168 @@ public boolean areThereResponsesForQuestion(UUID questionId) {
+
+        switch (giverType) {
+        case INSTRUCTORS:
+        case SELF:
+            for (Instructor instructor : roster.getInstructors()) {
+                numberOfRecipients =
+                        fqLogic.getRecipientsOfQuestion(question, instructor, null, roster).size();
+                responses = getFeedbackResponsesFromGiverForQuestion(question.getId(), instructor.getEmail());
+                updates.addAll(FeedbackRankRecipientsResponseDetails
+                        .getUpdatedResponsesForRankRecipientQuestionsMigrated(responses, numberOfRecipients));
+            }
+            break;
+        case TEAMS:
+        case TEAMS_IN_SAME_SECTION:
+            Student firstMemberOfTeam;
+            String team;
+            Map<String, List<Student>> teams = roster.getTeamToMembersTable();
+            for (Map.Entry<String, List<Student>> entry : teams.entrySet()) {
+                team = entry.getKey();
+                firstMemberOfTeam = entry.getValue().get(0);
+                numberOfRecipients =
+                        fqLogic.getRecipientsOfQuestion(question, null, firstMemberOfTeam, roster).size();
+                responses =
+                        getFeedbackResponsesFromTeamForQuestion(
+                                question.getId(), question.getCourseId(), team, roster);
+                updates.addAll(FeedbackRankRecipientsResponseDetails
+                        .getUpdatedResponsesForRankRecipientQuestionsMigrated(responses, numberOfRecipients));
+            }
+            break;
+        default:
+            for (Student student : roster.getStudents()) {
+                numberOfRecipients =
+                        fqLogic.getRecipientsOfQuestion(question, null, student, roster).size();
+                responses = getFeedbackResponsesFromGiverForQuestion(question.getId(), student.getEmail());
+                updates.addAll(FeedbackRankRecipientsResponseDetails
+                        .getUpdatedResponsesForRankRecipientQuestionsMigrated(responses, numberOfRecipients));
+            }
+            break;
+        }
+
+        for (FeedbackResponse update : updates) {
+            try {
+                updateFeedbackResponse(update);
+            } catch (EntityAlreadyExistsException | EntityDoesNotExistException | InvalidParametersException e) {
+                assert false : "Exception occurred when updating responses after deleting students.";
+            }
+        }
+    }
+
+    /**
+     * Updates a feedback response.
+     *
+     * <p>If the giver/recipient field is changed, the response is updated by recreating the response
+     * as question-giver-recipient is the primary key.
+     */
+    void updateFeedbackResponse(FeedbackResponse feedbackResponse)
+            throws EntityAlreadyExistsException, EntityDoesNotExistException, InvalidParametersException {
+        FeedbackResponse oldResponse = frDb.getFeedbackResponse(feedbackResponse.getId());
```

Comment:
Remember, we have mutated the existing feedback response. There is no concept of new and old feedback response anymore. The old and new feedback response refer to the same object.There should be no need to delete and recreate any responses.

With this in mind, can you look through your methods again. There is a lot of room for simplification here.

Reply:
Okay will look through everything!